config/crd/bases/elasticache.services.k8s.aws_users.yaml

```diff
@@ -0,0 +1,164 @@
+
+---
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  annotations:
+    controller-gen.kubebuilder.io/version: v0.4.0
+  creationTimestamp: null
+  name: users.elasticache.services.k8s.aws
+spec:
+  group: elasticache.services.k8s.aws
+  names:
+    kind: User
+    listKind: UserList
+    plural: users
+    singular: user
+  scope: Namespaced
+  versions:
+  - name: v1alpha1
+    schema:
+      openAPIV3Schema:
+        description: User is the Schema for the Users API
+        properties:
+          apiVersion:
+            description: 'APIVersion defines the versioned schema of this representation
+              of an object. Servers should convert recognized schemas to the latest
+              internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources'
+            type: string
+          kind:
+            description: 'Kind is a string value representing the REST resource this
+              object represents. Servers may infer this from the endpoint the client
+              submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
+            type: string
+          metadata:
+            type: object
+          spec:
+            description: UserSpec defines the desired state of User
+            properties:
+              accessString:
+                description: Access permissions string used for this user.
+                type: string
+              engine:
+                description: The current supported value is Redis.
+                type: string
+              noPasswordRequired:
+                description: Indicates a password is not required for this user.
+                type: boolean
+              passwords:
+                description: Passwords used for this user. You can create up to two
+                  passwords for each user.
+                items:
+                  type: string
+                type: array
```

Comment:
@kumargauravsharma  see the PR description... @echen-98 identified this as being a problem.

Since `Passwords` is a slice of string and `*ackv1alpha1.SecretKeyReference` refers to a single secret key, not multiple things, we have a mismatch here.

If you mark `User.spec.passwords` as `is_secret`, you will essentially need to write custom code for handling all of User's newCreateBuildRequest and setOutput code points. This custom code will need to translate between a single string value stored in the SecretKeyReference and the multiple strings needed in the `Passwords` field of the CreateUserInput shape... You will need to deal with the whole "up to two passwords" constraint manually in custom code and have the user place the passwords in a comma-delimited string in the SecretKeyReference storage.

Alternately, you could set the `User.spec.passwords` and `User.spec.noPasswordRequired` fields as `is_ignored` in the generator config and set `noPasswordRequired` to always be `true` in the CreateUserInput shape...

Reply:
@jaypipes Thanks for pointing that out. I was initially considering something closer to the latter option and leaving support for the `passwords` field for later. As a result I wasn't aware of this additional challenge. 

At the same time the `passwords` are an important part of the User APIs, so I'll look into it further

Reply:
@jaypipes thanks for your inputs.
My comment was to have it as array (slice) of K8s Secret type. 
Thus the generated Spec code would have this property as:
```
Passwords []*ackv1alpha1.SecretKeyReference `json:"passwords,omitempty"`
```
and the corresponding crd would look like:
```
         passwords:
             description: Passwords used for this user. You can create up to two passwords for each user.
             type: array
             items:
                properties:
                  key:
                    description: Key is the key within the secret
                    type: string
                  name:
                    description: Name is unique within a namespace to reference a
                      secret resource.
                    type: string
                  namespace:
                    description: Namespace defines the space within which the secret
                      name must be unique.
                    type: string
                required:
                - key
                type: object
```
With this, there is no need to store multiple password strings into single K8s secret field.
Rather, each password in slice will have its corresponding K8s secret field for value.

Reply:
We considered not to exclude password fields from spec.

Reply:
@kumargauravsharma sounds good, I will follow up with another revision and then keep the support for the `passwords` field as a follow-up. 

I'll probably also rename the PR and commit to say "partial support" to convey this.

Reply:
PR: https://github.com/aws-controllers-k8s/code-generator/pull/87 adds support for K8s Secret for slice types.